src/LanguageServer/Protocol/Handler/Formatting/FormatDocumentOnTypeHandler.cs

```diff
@@ -48,14 +51,27 @@ public FormatDocumentOnTypeHandler(IGlobalOptionService globalOptions)
             return [];
         }
 
-        var position = await document.GetPositionFromLinePositionAsync(ProtocolConversions.PositionToLinePosition(request.Position), cancellationToken).ConfigureAwait(false);
+        var linePosition = ProtocolConversions.PositionToLinePosition(request.Position);
+        var position = await document.GetPositionFromLinePositionAsync(linePosition, cancellationToken).ConfigureAwait(false);
 
         var formattingService = document.Project.Services.GetRequiredService<ISyntaxFormattingService>();
         var documentSyntax = await ParsedDocument.CreateAsync(document, cancellationToken).ConfigureAwait(false);
-
-        if (!formattingService.ShouldFormatOnTypedCharacter(documentSyntax, request.Character[0], position, cancellationToken))
+        var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+        // The formatting service expects that the position is inside the token span associated with the typed character, but
+        // in VSCode this is not always the case - the position the client gives us is not necessarily the position of the typed character.
+        // For example typing , the client may automatically
+        //   1.  '\n' - the client inserts indentation (to get '\n    ')
+        //   2.  '{' - the client inserts '}' (to get '{}')
+        //
+        // When the formatter calls root.FindToken, it may return a token different from what triggered the on type formatting, which causes us
+        // to format way more than we want, depending on exactly where the position ends up.
+        // Here we do our best to adjust the position back to the typed char location.
+        if (text[position - 1] != request.Character[0])
```

Comment:
this causes the formatter to not attempt to move braces when you hit enter between them, e.g.
```
void M() {<enter>}
```
will not get formatted to
```
void M()
{
    
}
```
as part of format on type.  This matches the behavior of VS format on type.